### ISSUE TITLE
Eliminate '...' var args from Lua code where possible

### ISSUE THREAD

**Issue by Gerold103** [tarantool/vshard#266]
It appears functions taking `...` can't be JITed if called not in a loop.

**Comment by Serpentian** [comment]
I see major performance degradation after the patch, which get rids of var args (majorly on router):

<details><summary>Details</summary>
<p>

```diff
diff --git a/test/upgrade/upgrade.result b/test/upgrade/upgrade.result
index e0c9469b..78942096 100644
--- a/test/upgrade/upgrade.result
+++ b/test/upgrade/upgrade.result
@@ -186,6 +186,7 @@ vshard.storage._call('test_api', 1, 2, 3)
  | - 1
  | - 2
  | - 3
+ | - null
  | ...
 
 test_run:switch('storage_1_b')
diff --git a/vshard/router/init.lua b/vshard/router/init.lua
index 1540509a..738506ec 100644
--- a/vshard/router/init.lua
+++ b/vshard/router/init.lua
@@ -722,27 +722,32 @@ end
 --
 -- Wrappers for router_call with preset mode.
 --
-local function router_callro(router, bucket_id, ...)
-    return router_call_impl(router, bucket_id, 'read', false, false, ...)
+local function router_callro(router, bucket_id, func, args, opts)
+    return router_call_impl(router, bucket_id, 'read', false, false,
+                            func, args, opts)
 end
 
-local function router_callbro(router, bucket_id, ...)
-    return router_call_impl(router, bucket_id, 'read', false, true, ...)
+local function router_callbro(router, bucket_id, func, args, opts)
+    return router_call_impl(router, bucket_id, 'read', false, true,
+                            func, args, opts)
 end
 
-local function router_callrw(router, bucket_id, ...)
-    return router_call_impl(router, bucket_id, 'write', false, false, ...)
+local function router_callrw(router, bucket_id, func, args, opts)
+    return router_call_impl(router, bucket_id, 'write', false, false,
+                            func, args, opts)
 end
 
-local function router_callre(router, bucket_id, ...)
-    return router_call_impl(router, bucket_id, 'read', true, false, ...)
+local function router_callre(router, bucket_id, func, args, opts)
+    return router_call_impl(router, bucket_id, 'read', true, false,
+                            func, args, opts)
 end
 
-local function router_callbre(router, bucket_id, ...)
-    return router_call_impl(router, bucket_id, 'read', true, true, ...)
+local function router_callbre(router, bucket_id, func, args, opts)
+    return router_call_impl(router, bucket_id, 'read', true, true,
+                            func, args, opts)
 end
 
-local function router_call(router, bucket_id, opts, ...)
+local function router_call(router, bucket_id, opts, func, args, call_opts)
     local mode, prefer_replica, balance
     if opts then
         if type(opts) == 'string' then
@@ -758,7 +763,7 @@ local function router_call(router, bucket_id, opts, ...)
         mode = 'write'
     end
     return router_call_impl(router, bucket_id, mode, prefer_replica, balance,
-                            ...)
+                            func, args, call_opts)
 end
 
 --
@@ -1657,15 +1662,16 @@ end
 -- Public API protection
 --------------------------------------------------------------------------------
 
-local function router_api_call_safe(func, router, ...)
-    return func(router, ...)
+local function router_api_call_safe(func, router, arg1, arg2, arg3, arg4, arg5)
+    return func(router, arg1, arg2, arg3, arg4, arg5)
 end
 
 --
 -- Unsafe proxy is loaded with protections. But it is used rarely and only in
 -- the beginning of instance's lifetime.
 --
-local function router_api_call_unsafe(func, router, ...)
+local function router_api_call_unsafe(func, router, arg1, arg2,
+                                      arg3, arg4, arg5)
     -- Router can be started on instance with unconfigured box.cfg.
     if not router.is_configured then
         local msg = 'router is not configured'
@@ -1676,12 +1682,12 @@ local function router_api_call_unsafe(func, router, ...)
         return error(lerror.vshard(lerror.code.ROUTER_IS_DISABLED, msg))
     end
     router.api_call_cache = router_api_call_safe
-    return func(router, ...)
+    return func(router, arg1, arg2, arg3, arg4, arg5)
 end
 
 local function router_make_api(func)
-    return function(router, ...)
-        return router.api_call_cache(func, router, ...)
+    return function(router, arg1, arg2, arg3, arg4, arg5)
+        return router.api_call_cache(func, router, arg1, arg2, arg3, arg4, arg5)
     end
 end
 
@@ -1739,8 +1745,8 @@ local module = {}
 -- This metatable bypasses calls to a module to the static_router.
 local module_mt = {__index = {}}
 for method_name, method in pairs(router_mt.__index) do
-    module_mt.__index[method_name] = function(...)
-        return method(M.static_router, ...)
+    module_mt.__index[method_name] = function(arg1, arg2, arg3, arg4, arg5)
+        return method(M.static_router, arg1, arg2, arg3, arg4, arg5)
     end
 end
 
diff --git a/vshard/storage/init.lua b/vshard/storage/init.lua
index 79b603c7..0147b7b9 100644
--- a/vshard/storage/init.lua
+++ b/vshard/storage/init.lua
@@ -3361,8 +3361,8 @@ service_call_api = setmetatable({
     return res
 end})
 
-local function service_call(service_name, ...)
-    return service_call_api[service_name](...)
+local function service_call(service_name, arg1, arg2, arg3, arg4)
+    return service_call_api[service_name](arg1, arg2, arg3, arg4)
 end
 
 --------------------------------------------------------------------------------
diff --git a/vshard/util.lua b/vshard/util.lua
index af49d97e..9d95280a 100644
--- a/vshard/util.lua
+++ b/vshard/util.lua
@@ -175,12 +175,12 @@ end
 -- @retval Wrapped function.
 --
 local function generate_self_checker(obj_name, func_name, mt, func)
-    return function (self, ...)
+    return function (self, arg1, arg2, arg3, arg4, arg5)
         if getmetatable(self) ~= mt then
             local fmt = 'Use %s:%s(...) instead of %s.%s(...)'
             error(string.format(fmt, obj_name, func_name, obj_name, func_name))
         end
-        return func(self, ...)
+        return func(self, arg1, arg2, arg3, arg4, arg5)
     end
 end
```

</p>
</details> 

Before patch: 176740.3000
After: 172541.0000 (-2.5%)

Let's ask @Buristan, whether there're any problems with JITing of the '...' var args

**Comment by Buristan** [comment]
The problem in the first place is that most of the functions (called on your workload) in the vshard can't be recorded due to various issues.
Hence, the removig of VARARG tail calls may cause the performance degradation.

The following patch on the top of tarantool/vshard-ee#51 enables the dumping of the traces to the corresponding files from routers:
```diff
diff --git a/example/router.lua b/example/router.lua
index 6817137..adf674d 100755
--- a/example/router.lua
+++ b/example/router.lua
@@ -6,6 +6,8 @@ fiber = require('fiber')
 local fio = require('fio')
 local NAME = fio.basename(arg[0], '.lua')
 
+require("jit.dump").start("ibxT", "/tmp/vshard_"..NAME..".jdump")
+
 -- Check if we are running under test-run
 if os.getenv('ADMIN') then
     test_run = require('test_run').new()
```

<details>

<summary> Analysis of the jit.dump output for router </summary>

All my observations are based on the content of the `jit.dump` result:

```
grep NYI -R vshard_router_1.jdump | perl -lnE 'say s/(?:.*)(NYI.*)/$1/r' | sort | uniq -c | sort -n
      5 NYI: bytecode 51 ; FNEW
      7 NYI: bytecode 71 ; VARG
      9 NYI: bytecode 70 ; ITERN
     78 NYI: bytecode 72 ; ISNEXT
```
The most often reason for recording failure is recording of `pairs()` (`next()`), not VARARG functions.
See also, tarantool/tarantool#6475.

We may see the code chunks that are tried to be recorded most (for the router):
```
grep start.*init.lua -R vshard_router_1.jdump | perl -lnE 'say s/(TRACE) \d\d\d (.*)/$1 $2/r' | sort | uniq -c | sort -n
      1 ---- TRACE start 132/stitch init.lua:383
      1 ---- TRACE start 133/stitch init.lua:408
      1 ---- TRACE start init.lua:113
      1 ---- TRACE start init.lua:156
      1 ---- TRACE start init.lua:365
      3 ---- TRACE start init.lua:235
      3 ---- TRACE start init.lua:274
      5 ---- TRACE start 105/0 init.lua:157
      5 ---- TRACE start 113/0 init.lua:114
     11 ---- TRACE start init.lua:1668
     11 ---- TRACE start init.lua:1692
     11 ---- TRACE start init.lua:1751
     11 ---- TRACE start init.lua:590
     11 ---- TRACE start init.lua:630
     11 ---- TRACE start init.lua:728
```

After all, several functions are blacklisted and start blacklisting cancer:
```
grep init.lua.*blacklisted -R vshard_router_1.jdump
---- TRACE 123 abort init.lua:1668 -- blacklisted
---- TRACE 123 abort init.lua:590 -- blacklisted
grep blacklisted -R vshard_router_1.jdump | perl -lnE 'say s/(TRACE) \d\d\d (.*)/$1 $2/r' | sort | uniq -c | sort -n
      1 ---- TRACE abort init.lua:1668 -- blacklisted
      1 ---- TRACE abort init.lua:590 -- blacklisted
      1 ---- TRACE abort internal.utils.lua:52 -- blacklisted
      3 ---- TRACE abort replicaset.lua:874 -- blacklisted
      3 ---- TRACE abort table.lua:42 -- blacklisted
      6 ---- TRACE abort replicaset.lua:389 -- blacklisted
     15 ---- TRACE abort replicaset.lua:647 -- blacklisted
     24 ---- TRACE abort internal.pairs.lua:7 -- blacklisted
     46 ---- TRACE abort fiber.lua:62 -- blacklisted
```

The `fiber.clock()` function is blacklisted due to the same reason as for `tuple_bless()` compilation. For the details, see:
* https://lists.tarantool.org/pipermail/tarantool-patches/2021-October/026241.html
* tarantool/tarantool@9649897
* https://github.com/tarantool/tarantool/issues/5896#issuecomment-2671297486

This may be patched at the Tarantool side, as well.

The `replica_call()`(`replicaset.lua:647`) is blacklisted due to several reasons:
1) Possible blacklisting of `fiber.clock()`, see above.
2) The trace ends up in the `remote_methods:reload_schema()` which calls `check_remote_arg(self, 'reload_schema')` which uses `pairs()`.
```
---- TRACE 105 start replicaset.lua:647
...
0000  . . . . FUNCC               ; pairs
0019  . . . ISNEXT   5 => 0119
---- TRACE 105 abort internal.utils.lua:60 -- NYI: bytecode 72
```

</details>

---

TLDR:
1) We may try to enable JIT-ing partially by avoiding the tail call to FFI in `fiber.clock()`.
2) Avoid such usage of `pairs()` on the critical path or backport the corresponding patches in the scope of tarantool/tarantool#6475 (note that pairs recording may not give so much speedup to the whole workload).
